src/mgr/DaemonServer.cc

```diff
@@ -1148,6 +1148,201 @@ void DaemonServer::_maximize_ok_to_stop_set(
   }
 }
 
+void DaemonServer::_update_upgraded_osds(
+  const std::set<int>& orig_osds,
+  const std::set<int>& to_upgrade,
+  const std::set<int>& upgraded,
+  const std::set<int>& version_unknown,
+  upgrade_osd_report *report)
+{
+  // reset output
+  *report = upgrade_osd_report();
+  report->osds = orig_osds;
+  report->ok_upgrade = to_upgrade;
+  report->ok_upgraded = upgraded;
+  report->bad_no_version = version_unknown;
+}
+
+void DaemonServer::_maximize_ok_to_upgrade_set(
+  const std::set<int>& orig_osds,
+  unsigned max,
+  const OSDMap& osdmap,
+  const PGMap& pgmap,
+  std::string_view ceph_version_new,
+  upgrade_osd_report *out_osd_report,
+  offline_pg_report *out_pg_report)
+{
+  std::set<int> to_upgrade;
+  std::set<int> upgraded;
+  std::set<int> version_unknown;
+
+  dout(20) << "orig_osds " << orig_osds
+           << " new ceph_version " << ceph_version_new << dendl;
+  // Filter osds not yet running the new ceph_version.
+  // Limit the check for safe upgrade to only the set
+  // of OSDs that are still running the older version.
+  for (auto& osd : orig_osds) {
+    auto osd_id = "osd." + std::to_string(osd);
+    auto ver = get_osd_metadata("ceph_version_short", osd_id);
+    if (ver.has_value()) {
+      if (*ver != ceph_version_new) {
```

Comment:
_Not equal_. Is the intention to handle downgrades the same way?

Reply:
Yes, as long as the version string passed to the command is correct and is different from the version the OSD is running on, this check will work for downgrades as well.